clients/drcachesim/tools/common/decode_cache.h

```diff
@@ -0,0 +1,390 @@
+ * information.
+ */
+
+#ifndef _DECODE_CACHE_H_
+#define _DECODE_CACHE_H_ 1
+
+#include "memref.h"
+#include "tracer/raw2trace_shared.h"
+
+#include <cstring>
+#include <mutex>
+#include <unordered_map>
+
+namespace dynamorio {
+namespace drmemtrace {
+
+/**
+ * Base class for storing instruction decode info. Users should sub-class this
+ * base class and implement set_decode_info_derived() to derive and store the decode
+ * info they need.
+ */
+class decode_info_base_t {
+public:
+    /**
+     * Sets the decode info for the provided \p instr which was allocated using the
+     * provided \p dcontext for the provided \p memref_instr. This is done using
+     * the set_decode_info_derived() provided by the derived class. Additionally,
+     * this does other required bookkeeping.
+     */
+    void
+    set_decode_info(void *dcontext,
+                    const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                    instr_t *instr);
+
+    /**
+     * Indicates whether the decode info stored in this object is valid. It won't be
+     * valid if the object is default-constructed without a subsequent
+     * set_decode_info() call. When used with \p decode_cache_t, this indicates
+     * that an invalid instruction was observed at some pc.
+     */
+    bool
+    is_valid() const;
+
+private:
+    /**
+     * Sets the decoding info fields as required by the derived class, based on the
+     * provided instr_t which was allocated using the provided opaque \p dcontext
+     * for the provided \p memref_instr. Derived classes must implement this virtual
+     * function. Note that this cannot be invoked directly as it is private, but
+     * only through set_decode_info() which does other required bookkeeping.
+     *
+     * This is meant for use with \p decode_cache_t, which will invoke
+     * set_decode_info() for each new decoded instruction.
+     *
+     * The responsibility for invoking instr_destroy() on the provided \p instr
+     * lies with this \p decode_info_base_t object, unless
+     * \p decode_cache_t was constructed with \p persist_decoded_instrs_
+     * set to false, in which case no heap allocation takes place.
+     */
+    virtual void
+    set_decode_info_derived(void *dcontext,
+                            const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                            instr_t *instr) = 0;
+
+    bool is_valid_ = false;
+};
+
+/**
+ * Decode info including the full decoded instr_t. This should be used with an
+ * \p decode_cache_t constructed with \p persist_decoded_instrs_ set to
+ * true.
+ */
+class instr_decode_info_t : public decode_info_base_t {
+public:
+    ~instr_decode_info_t();
+    instr_t *
+    get_decoded_instr();
+
+private:
+    void
+    set_decode_info_derived(void *dcontext,
+                            const dynamorio::drmemtrace::_memref_instr_t &memref_instr,
+                            instr_t *instr) override;
```

Comment:
Hmm but from above the base class owns this instr: so this class has to clone it (have not looked at .cpp files yet)?

Reply:
I suppose this is resolved now? The derived class does take ownership if used with a decode_cache_t where persist_decoded_instrs_ is set to true.

Reply:
The base class wouldn't know how the user intends to use the derived class (whether they want to use it with a decode_cache_t with persist_decoded_instrs set to true).